Meta: Clearly document current guidance for changing existing PEPs

CAM-Gerlach opened the pull request:
As has come up on many issues and PRs submitted here (e.g. #2151, #2199. #2265, #2318. #2332 being but a few recent examples off the top of my head), it would be of benefit to provide clear unambiguous up-front guidance on whether, when and where to propose changes to existing PEPs. Therefore, this PR adds this near the top of the Contributing Guide (which is exposed fairly prominently by the GitHub UI), to explictly explain the established, mostly unwritten practice here.

Specifically, it:

* [x] Clarifies that non-Draft/Active PEPs are historical documents that should generally be left alone
* [x] Briefly describes what types of changes are appropriate for each major type/status of PEP
* [x] Mentions that issues/PRs here are not usually a good place for substantive debate on PEP content, and redirects users to appropriate discussion venues and/or proposing a new PEP if they'd like to see substantive changes
* [x] Notes that the SC should be notified and any PEP-specific procedures followed when changing governance PEP (to avoid any future uncertainty like on #2346)
* [x] Retains and emphasizes the specific discouragement of mass-spellcheck/etc, PRs changing old PEPs (originally added in #2338 as requested in #2151)

Also, I added a sentence to the "Commit messages and PR titles" section clarifying prefix usage for non-PEP PR titles/commit messages, as that has been the source of some confusion lately by multiple PEP editors (e.g. on  #2364 and #2375 )

gvanrossum reviewed (commented):
I think the bar for Active PEPs is higher. Maybe ask @warsaw .

JelleZijlstra commented on CONTRIBUTING.rst:
```diff
@@ -18,12 +18,71 @@ first in an appropriate venue, drafting a PEP and gathering feedback, and
+Contributing changes to existing PEPs
+-------------------------------------
+
+In general, PEPs that are not marked as Draft or Active are usually considered
```
Remove one of "In general" and "usually". Come to think of it, probably remove both.

CAM-Gerlach replied:
I removed the latter; I wanted to hedge a bit here due to provisional PEPs, some PEPs that are de-facto specifications and updated from time to time, and other edge cases—though there has been a _general_ push to reduce that over time and move most of that elsewhere to more suitable platforms.

JelleZijlstra commented on CONTRIBUTING.rst:
```diff
@@ -18,12 +18,71 @@ first in an appropriate venue, drafting a PEP and gathering feedback, and
+In general, PEPs that are not marked as Draft or Active are usually considered
+to be historical documents rather than documentation or living specifications.
+To propose a substantially revised version of the original ideas,
+you should generally consider opening a new PEP instead.
```
Though not every change to something specified in a PEP will itself need a PEP to change.

CAM-Gerlach replied:
Yeah—if this sounds a bit out of place here, that's because it was originally part of the first bullet (Deferred/Withdrawn/Rejected), was more about do-overs for old PEPs. I realized it could apply, in somewhat modified form, to Final and Accepted PEPs too, so I moved it here.

In any case, to address this I completely re-wrote this paragraph to strengthn the conditions for a new PEP, add more alternatives, integrate the sentences better and add "Provisional" to "Active" and "Draft" above.

JelleZijlstra commented on CONTRIBUTING.rst:
```diff
@@ -18,12 +18,71 @@ first in an appropriate venue, drafting a PEP and gathering feedback, and
+As such, make sure to consider the type and status of the PEP(s)
+in question before proposing changes:
+
+* **Rejected**, **Deferred**, **Withdrawn** and **Superceded** PEPs should
```
Not sure about Deferred or Withdrawn. Deferred PEPs could see updates to bring them up to date; and a Withdrawn PEP could be prepared for a new shot.

CAM-Gerlach replied:
I was initially going to say something like this here (see above), but my general impression (which might be off base) was that for old PEPs that got enough out of step with modern Python to earn such a status (as the great majority of Deferred and Withdrawn PEPs are, since only PEPs >=several years old seem to have either status, aside from PEPs withdrawn for other reasons) it was usually recommended to start fresh with a new PEP rather than completely re-writing an old one. But that might not be an accurate impression, since I can't point to somewhere in PEP 1 where it says that (as opposed to for Rejected PEPs).

In any case, I've revised this bullet point to offer revival as a possibility.

JelleZijlstra commented on CONTRIBUTING.rst:
```diff
@@ -18,12 +18,71 @@ first in an appropriate venue, drafting a PEP and gathering feedback, and
+  intended.
+
+* **Final** PEPs are considered complete; while small fixes to clearly
+  unintentional errata in the document that meaningfully impair understanding
```
```suggestion
  unintentional errors in the document that meaningfully impair understanding
```
Errata fix errors

CAM-Gerlach replied:
Gotcha, corrected

JelleZijlstra commented on CONTRIBUTING.rst:
```diff
@@ -18,12 +18,71 @@ first in an appropriate venue, drafting a PEP and gathering feedback, and
+  at the discretion of the author. Copyediting, formatting and proofreading
+  can be opened here as pull requests, while more substantive content
+  changes should generally be first proposed on the PEP's discussion thread
+  listed in its ``Discussions-To`` header.
```
We should also mention that once the PEP has been submitted for pronouncement to the SC or delegate, further changes could delay the process.

CAM-Gerlach replied:
Good point; I revised this paragraph accordingly

hugovk commented on CONTRIBUTING.rst:
```diff
@@ -18,12 +18,72 @@ first in an appropriate venue, drafting a PEP and gathering feedback, and
+  may be considered, substantive changes to the content generally will not.
+
+* **Accepted** PEPs may be updated to reflect minor details of the final
+  implementation, but significant changes to the normative specification
```
Is there a plain English synonym for "normative" that can be used? I had to look it up :)

JelleZijlstra replied:
We can just remove "normative".

CAM-Gerlach replied:
I just removed it; its an extra word and isn't critical to meaning here.

EDIT: Seems like I forgot to both actually push my changes from many hours ago, and submit this reply/resolve this comment. Oops, my fault.

CAM-Gerlach commented:
I've edited down the text to reduce length and prescriptiveness, particularly in the bullets summarizing the guidelines for each type of PEP (which was the actual sorta-prescriptive part). In total, this section now only adds 49 lines (including blanks) to to the 178 line document, or just above  1/4 of the total length, while removing 8 lines from the spellcheck section of text that was moved to this one, for a delta of only +41 lines, or 2/9 the total (and 4 of _those_ are just URLs that don't render); by contrast, just the linting sections are 75 lines.

This seems fairly appropriate, since this section is framed in terms of guidelines rather than hard rules, not only discusses "making changes to old PEPs", but also a number of directly related issues that have come up recently or been otherwise requested here, including:

* What should be brought up here versus on the PEP's discussion thread
* Discouraging "omnibus" mess spellcheck/etc. PRs
* When to open a new PEP versus revising an old one
* To avoid major modifications to PEPs after they are submitted for approval
* SC assent for changes to governance PEPs
* Where potential contributors can reach out to clarify specific situations further

In total, these issues _do_ come up fairly frequently (see examples above) here and elsewhere, and while some users still may miss it anyway (though GitHub's UI does its best), it gives us a section to link to when it does (as others have requested), and contributors tend to react much better to a documented guideline they missed, than an unwritten rule they are only told after putting effort into a PR.

Overall, I agree we shouldn't be overly prescriptive, but since we have these mostly unwritten rules that we're enforcing (somewhat inconsistently), I think it would be best to document them and agree on at least a set of guidelines so that contributors and ourselves have a common baseline understanding of them and know (how) to reach out to clarify for specific situations.

JelleZijlstra commented on CONTRIBUTING.rst:
```diff
@@ -18,12 +18,72 @@ first in an appropriate venue, drafting a PEP and gathering feedback, and
 When adding or modifying a PEP, please always include the PEP number in the
 commit summary and pull request title.
 For example, ``PEP NNN: <summary of changes>``.
+Likewise, prefix rendering infrastructure changes with ``Infra:``,
```
I think we can skip this sentence for brevity. It's not that important that the prefixes are consistent (though I agree with you that we should avoid obscure abbreviations like "PRS").

CAM-Gerlach replied:
We can, though I don't think it hurts to be provide some basic guidance here, since a lack of such was how we got cryptic ad-hoc initialisms like "PRS" in the first place, and its helpful to see at a glance the general high-level topic of a commit/PR without having to parse the full title/commit message.

CAM-Gerlach replied:
We could also drop the "Lint" prefix and just have "Infra" and "Meta", if that would simplify things.

hugovk replied:
"Lint", "Infra" and "Meta" are good.

I'm also confused about "PRS", what does it stand for? It's very difficult to search for :)

CAM-Gerlach replied:
Seems like opinions were mixed about specifying particular prefixes (@hugovk and me :+1: , @JelleZijlstra and @AA-Turner :+1: ), but us three plus Guido all found "PRS" very confusing, so I dropped this sentence for now on the condition that we all avoid bespoke initialisms like that going forward. Deal?

> I'm also confused about "PRS", what does it stand for? It's very difficult to search for :)

AFAIK @AA-Turner made it up, apparently (per his responses to Guido, Jelle, and me who all asked the same question). I believe it was intended to stand for "PEP Rendering System".

AA-Turner commented:
I've re-read the proposed changes & my suggestion for the entire section is:

----

## Contributing changes to existing PEPs

PEPs serve as a record for why a change was made, including a detailed specification for the change. This means that in general, PEPs are considered to be historical documents. Edits to existing PEPs may be accepted, although we encourage you to contact the PEP author(s) or gather consensus for the update first. PEP 1 [contains more guidance](https://python.github.io/peps/pep-0001/#pep-maintenance).

----

I appreciate this is rather different from the proposal here -- my rationale is that PEP 1 already contains text on this, and given that the CONTRIBUTING.rst document is non-normative, I don't want to make promises we can't keep.

A

CAM-Gerlach commented:
@AA-Turner I agree with your comments overall; it doesn't make sense to partially duplicate the relevant section in PEP 1, while not offering or directly referring to explicitly normative guidance, and it would be better to more clearly separate the purpose of the content in each, while referencing the other for the rest, and reducing the total length.

I do think it makes sense to retain some existing content unique to and more appropriate for the Contributing Guide there, including specific guidance on what to make as PRs (copyediting/proofreading draft PEPs), what to direct to the Discussions-To thread (substantive changes) and what to avoid (e.g. mass typo correction PRs, which was already there before per your own request); and mentioning resources to reach out to if unsure. Also, there are a few more substantive items not currently mentioned in PEP 1, including SC approval of changes to governance PEPs, holding changes to PEPs during final review and resolution, and reviving Deferred/Withdrawn PEPs.

Therefore, I went ahead with this, retaining the former existing helpful items in the Contributing Guide, revising PEP 1 to reflect those missing bits, and cutting everything else.

gvanrossum commented on CONTRIBUTING.rst:
```diff
@@ -18,12 +18,37 @@ first in an appropriate venue, drafting a PEP and gathering feedback, and
+Contributing changes to existing PEPs
+-------------------------------------
+
+In general, most non-Draft/Active PEPs are considered to be historical
+documents rather than living specifications or documentation. Major changes to
+their core content usually require a new PEP, while smaller modifications may
+or may not be appropriate, depending on the PEP's status. See `PEP 1
+<https://www.python.org/dev/peps/pep-0001/#pep-maintenance>`__ for more.
```
I understand this is just a summary, but there's such a big divide between Draft and Active PEPs (SC approval comes to mind :-) that it's wrong to describe them as more or less the change here.

gvanrossum replied:
(Sorry, I confused Active with Accepted. Never mind then.)

gvanrossum commented on CONTRIBUTING.rst:
```diff
@@ -18,12 +18,37 @@ first in an appropriate venue, drafting a PEP and gathering feedback, and
+or may not be appropriate, depending on the PEP's status. See `PEP 1
+<https://www.python.org/dev/peps/pep-0001/#pep-maintenance>`__ for more.
+
+Copyediting and proofreading Draft and Active PEPs (including fixing spelling
+and formatting issues) is welcomed, and can be done via pull request to this
```
I think the author should be allowed to have total control. In the end the PEP repo is a convenient place where authors place their proposals to be discussed by the community. Authors may appreciate help, but PEPs, unlike general documentation, *have* authors, and the process is geared towards their needs.

CAM-Gerlach replied:
I originally included an explicit conditional on this line, "at the discretion of the author(s)", but eliminated it in response to the request to make the more concise. I added back appropriate caveats both places this is referred to.

(As a sidenote, there are _some_ limits to what is an author's discretion over a PEP editor, when push comes to shove—insensitive or inappropriate language, syntax/technical issues, major breaches of PEP 1/12, etc. But not need to delve into that here.)

gvanrossum commented on pep-0001.txt:
```diff
@@ -457,19 +457,30 @@ if they are never meant to be completed.  E.g. :pep:`1` (this PEP).
-reached the Final state. Once a PEP has been completed, the Language and
-Standard Library References become the formal documentation of the expected
-behavior.
+In general, PEPs are no longer substantially modified after they have reached
+the Final, Rejected or Superseded state. Once a PEP has been completed,
```
This is also true (with only minor exceptions) for the Accepted state.

CAM-Gerlach replied:
Yeah; that was my impression and what the original wording of my Contributing Guide addition said, but when I moved it to PEP 1, the present wording there only described this for the "Final" state, and mentioned changes to "Accepted" PEPs on equal footing with "Provisional" ones. I've revised this and the passage below accordingly to reflect the current reality here.

CAM-Gerlach commented:
Oops, sorry, I accidentally dismissed your :heavy_check_mark: @JelleZijlstra — its been a hectic few days for me.

AA-Turner commented on pep-0001.txt:
```diff
@@ -457,19 +457,31 @@ if they are never meant to be completed.  E.g. :pep:`1` (this PEP).
 If changes based on implementation experience and user feedback are made to
-Standards track PEPs while in the Accepted or Provisional State, those changes
-should be noted in the PEP, such that the PEP accurately describes the state of
+Standards track PEPs while in the Provisional or (with SC approval) Accepted
```
Do we specify anywhere else that changes to Accepted PEPs require re-approval? If so, the `(with SC approval)` can be taken out. If not, please can this be rephrased to not have the parenthetical aside.

CAM-Gerlach replied:
Not that I could find, but its possible I missed it—this PEP is rather long and its organization sometimes results in things being different places. I was basing this off @gvanrossum 's reply here and what seems to be accepted (heh) practice.

We can do this, if you prefer:

```suggestion
Standards track PEPs while in the Provisional (or with SC approval, Accepted)
```

Off the top of my head, I'm not sure I can think of another straightforward way of avoiding this while still being explicit and unambitious that it only applies to `Accepted` and not `Provisional`, at least without rewriting this whole otherwise unchanged paragraph. Could you maybe suggest something, and/or explain the issue with the parenthetical?

AA-Turner replied:
> explain the issue with the parenthetical

A general aversion to parenthetical asides in formal writing. Suggestion below, though it is a re-write of the entire paragraph.

A

```rest
While implementing a Standards Track PEP, it may be desirable to alter the
behaviour described in the PEP due to user feedback or implementation 
experience. The PEP should be updated such that it accurately describes
the implementation at the point where it is marked Final. Changes to the
behaviour described in an Accepted PEP must also recieve approval from the 
`Steering Council`_.
```

CAM-Gerlach replied:
I'm not sure we need to completely re-write this paragraph from the original just to suit one person's personal aversion to parenthetical asides, but what do others think (@JelleZijlstra @hugovk )?

hugovk replied:
I think it's fine as is.

AA-Turner commented on pep-0001.txt:
```diff
@@ -457,19 +457,31 @@ if they are never meant to be completed.  E.g. :pep:`1` (this PEP).
+Active Informational and Process PEPs may be updated over time to reflect
+changes to development practices and other details. The precise process
+followed in these cases will depend on the nature and purpose of the PEP
+being updated. Substantive changes to governance PEPs must be reviewed by the
+Steering Council (by opening a `Steering Council issue`_).
```
Again, is this captured elsewhere to your satisfaction? If so the sentence starting "Substantive" can be removed.

CAM-Gerlach replied:
I did some searching (before and now) and I couldn't find anywhere else that explicitly mentioned this. With how long PEP 1 is, it is not impossible that I missed it, but at least it didn't appear to be anywhere I could think to find it, and this seems to be the most logical place to put it.

We could also modify this to say

```suggestion
being updated. Substantive changes to Active Process PEPs must be
reviewed by the Steering Council (by opening a `Steering Council issue`_).
```

since that more clearly defines what PEPs must have this approval, since "governance PEPs" are not so well defined.

To note, as I've noticed previously, PEP 13 is inexplicably listed as an Informational PEP rather than a Process PEP, despite it being the Meta/Process PEP of all Process PEPs as it defines the "process" by which the Python community governs itself and in turn decides on the PEP process, and is thus the most immutable of them all.

AA-Turner replied:
Your suggested wording looks good (although, I would actually interpret "governance PEPs" as just 13 and the 8000 series, so perhaps it would be useful to check what the Council care about enough to add to their workload!

A

CAM-Gerlach replied:
Any thoughts on this @brettcannon ?

encukou replied:
The SC *cannot* unilaterally change PEP 13. The procedure to change PEP 13 is described [in the PEP itself](https://peps.python.org/pep-0013/#changing-this-document). The same goes for the [PyPA governance PEP](https://peps.python.org/pep-0609/#updates-to-the-governance-specification-processes) for example.
IMO the previous wording is enough here, no addition about the SC is necessary.

CAM-Gerlach replied:
Right, but this is about Active Process PEPs in general, not anything specific to one PEP, which is what my suggestion above changes it to be more clear on (and, inexplicably, PEP 13 is listed as a mere _"Informational"_ PEP rather than a "Process" PEP, so technically it wouldn't apply to PEP 13 at all, albeit almost certainly due to an oversight). That said, the previous sentence says "The precise process followed in these cases will depend on the nature and purpose of the PEP being updated.", so if there's no need to explicitly specify this here, I'll just remove it.

```suggestion
in question.
```

To note, I originally included another clause specifying that beyond SC review, any guidance in the PEP in question [e.g. PEP 13, 609, etc] also needed to be followed when updating it, but I had elided it in the current version for brevity, due to concerns about the original proposed wording being too verbose and overly specific.

AA-Turner reviewed (commented):
A few comments:

A

CAM-Gerlach commented on CONTRIBUTING.rst:
```diff
@@ -18,12 +18,38 @@ first in an appropriate venue, drafting a PEP and gathering feedback, and
+
+If you're still unsure, we encourage you to reach out first before opening a
+PR here. For example, you could contact the PEP author(s), propose your idea in
+a discussion venue appropriate to the PEP (such as `Typing-SIG
+<https://mail.python.org/archives/list/typing-sig@python.org/>`__ for static
+typing, or `Packaging Discourse <https://discuss.python.org/c/packaging/>`__
+for packaging), or `open an issue <https://github.com/python/peps/issues>`__.
+
```
Your suggestion with fixed line breaks:

```suggestion
a discussion venue appropriate to the PEP's topic, or
`open an issue here <https://github.com/python/peps/issues>`__.
```

CAM-Gerlach commented on CONTRIBUTING.rst:
```diff
@@ -18,12 +18,38 @@ first in an appropriate venue, drafting a PEP and gathering feedback, and
-For example, ``PEP NNN: <summary of changes>``.
+When adding or modifying a PEP, please include the PEP number in the commit
+summary and pull request title. For example, ``PEP NNN: <summary of changes>``.
+Likewise, prefix rendering infrastructure changes with ``Infra:``, linting
+alterations with ``Lint:`` and other non-PEP meta changes, such as updates to
+the Readme/Contributing Guide, issue/PR template, etc., with ``Meta:``.
```
That was the convention I was following before, and I'm okay with that if others agree (provided we _do_ ensure that changes to specific PEPs are prefixed properly).

I don't think its necessary to be too prescriptive on this either (we have GitHub tags now anyway); this mostly was to avoid individual PEP editors coming up with their own cryptic, bespoke initialisms and confusing multiple other PEP editors, but that seems to be a pretty limited and specific problem that is better solved in other more appropriate ways.

```suggestion
```

CAM-Gerlach commented:
Just to note, this came up yet again in #2424 .

AA-Turner reviewed (approved):
Beyond the two outstanding discussions (which I've responded to), this now looks good -- I won't be around for the final reviews, so will leave in the capable hands of others.

A

gvanrossum commented:
I think if we’re now considering this document too formal for parenthetical remarks, we have taken a wrong turn somewhere (and not just because I’m too fond of them myself :-).

CAM-Gerlach commented:
I belatedly noticed there is also a (rather verbosely named) "Reporting PEP Bugs, or Submitting PEP Updates" section in PEP 1 that covers some similar ground to the section in the Contributing Guide, as well as some elements of what was added to the PEP Maintenance section.

As such, I've made some additional changes to this PR to focus each section more tightly, reduce duplication and further shorten the added length:

* Removed the "Draft" paragraph this PR added to the PEP Maintenance subsection of PEP Workflow to focus it more tightly on maintaining mature PEPs in the PEP lifecycle (as it did originally)
* Moved the content from the "Draft" paragraph to the (more concisely renamed) "Changing Existing PEPs" section, and updated/simplified the existing somewhat outdated content
* Further simplify the language and structure of the section added to the Contribtuing Guide
* Adding cross-linking between all three locations (PEP Maintenance, Changing Existing PEPs and Contrib guide)

Also, I rebased the PR after the implementation of PEP 676, and updated the added links accordingly.

CAM-Gerlach commented:
@python/pep-editors I've revised this one last time to further reduce duplication, cross-link the sections involved and slim down the total net additions. Any final feedback on this? Also, do you think its substantive enough to be worth submitting to the SC for assent before merging? As far as I'm aware, it only clarifies the existing _de jure_ and _de facto_ established practices, but I don't want to be too hasty about it. Thanks!

Rosuav commented:
LGTM. I think it's all just tightening up what's already been the case, so I don't think it should need SC approval, but it'd be nice to have at least one SC member sound off on that point.

CAM-Gerlach commented:
> LGTM. I think it's all just tightening up what's already been the case, so I don't think it should need SC approval, but it'd be nice to have at least one SC member sound off on that point.

:point_up: @brettcannon @encukou ? Also, your judgement on [this question](https://github.com/python/peps/pull/2378#discussion_r825495636) would be much appreciated.

CAM-Gerlach commented:
I'll wait another 24 hours in case anyone has further comments or objections, and if not, I'll go ahead and merge. Thanks everyone for all the great feedback that greatly shaped and improved the final result!

brettcannon commented on pep-0001.txt:
```diff
@@ -458,19 +458,25 @@ if they are never meant to be completed.  E.g. :pep:`1` (this PEP).
-Informational and Process PEPs may be updated over time to reflect changes
-to development practices and other details. The precise process followed in
-these cases will depend on the nature and purpose of the PEP being updated.
+Active Informational and Process PEPs may be updated over time to reflect
```
```suggestion
Active, Informational and Process PEPs may be updated over time to reflect
```

CAM-Gerlach replied:
Actually, the intended syntax is `Active (Informational and Process) PEPs`, not `(Active, Informational and Process) PEPs` as the above change would imply. Any suggestions to make that clearer (as that obviously wasn't what came across)? E.g.

```suggestion
Active-status Informational and Process PEPs may be updated over time to reflect
```

or

```suggestion
Informational and Process PEPs that are Active may be updated over time to reflect
```

or even as a literal parenthetical?

```suggestion
Active (Informational and Process) PEPs may be updated over time to reflect
```

hugovk replied:
The last one looks clearest to me.

CAM-Gerlach commented:
Its been another week and the one additional raised (from an approving review) was resolved, so I'll finally merge this now. Thanks again for all the great feedback!
